File: server/src/test/java/odm/clarity/woleh/places/util/PlaceNameNormalizerTest.java
package odm.clarity.woleh.places.util;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import odm.clarity.woleh.common.error.PlaceNameValidationException;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class PlaceNameNormalizerTest {

	private PlaceNameNormalizer normalizer;

	@BeforeEach
	void setUp() {
		normalizer = new PlaceNameNormalizer();
	}

	// ── PLACE_NAMES.md §5 canonical test vectors ───────────────────────────

	/** Vector 1: leading/trailing whitespace + case difference → equal. */
	@Test
	void vector1_leadingTrailingSpacesAndCaseAreEquivalent() {
		assertThat(normalizer.normalize("  Accra  "))
				.isEqualTo(normalizer.normalize("accra"));
	}

	/** Vector 2: decomposed accent (e + U+0301) equals precomposed (U+00E9) after NFC. */
	@Test
	void vector2_decomposedAndPrecomposedAccentAreEqual() {
		String decomposed = "e\u0301"; // e + combining acute accent
		String precomposed = "\u00E9"; // é precomposed
		assertThat(normalizer.normalize(decomposed))
				.isEqualTo(normalizer.normalize(precomposed));
	}

	/** Vector 3: multiple internal spaces collapse to one. */
	@Test
	void vector3_multipleInternalSpacesCollapsedToOne() {
		assertThat(normalizer.normalize("Main  St"))
				.isEqualTo(normalizer.normalize("Main St"));
	}

	// ── normalize: edge cases ─────────────────────────────────────────────

	@Test
	void normalize_emptyStringReturnsEmpty() {
		assertThat(normalizer.normalize("")).isEmpty();
	}

	@Test
	void normalize_nullReturnsEmpty() {
		assertThat(normalizer.normalize(null)).isEmpty();
	}

	@Test
	void normalize_blankStringReturnsEmpty() {
		assertThat(normalizer.normalize("   ")).isEmpty();
	}

	@Test
	void normalize_internalTabCollapsedToSpace() {
		assertThat(normalizer.normalize("Accra\tCentral"))
				.isEqualTo("accra central");
	}

	@Test
	void normalize_mixedCaseFolded() {
		assertThat(normalizer.normalize("CIRCLE")).isEqualTo("circle");
		assertThat(normalizer.normalize("Kaneshie")).isEqualTo("kaneshie");
	}

	@Test
	void normalize_exactOutputValues() {
		assertThat(normalizer.normalize("  Accra  ")).isEqualTo("accra");
		assertThat(normalizer.normalize("Main  St")).isEqualTo("main st");
	}

	@Test
	void normalize_isIdempotent() {
		String input = "  Kaneshie  Market  ";
		String once = normalizer.normalize(input);
		assertThat(normalizer.normalize(once)).isEqualTo(once);
	}

	@Test
	void normalize_mixedInternalWhitespace() {
		// Tab + multiple spaces between words
		assertThat(normalizer.normalize("Accra \t  Central"))
				.isEqualTo("accra central");
	}

	// ── validatePlaceName: valid inputs ──────────────────────────────────

	@Test
	void validatePlaceName_typicalNameDoesNotThrow() {
		normalizer.validatePlaceName("Accra Central");
	}

	@Test
	void validatePlaceName_exactlyMaxCodePointsAccepted() {
		String maxName = "a".repeat(PlaceNameNormalizer.MAX_CODE_POINTS);
		normalizer.validatePlaceName(maxName); // must not throw
	}

	// ── validatePlaceName: invalid inputs ────────────────────────────────

	@Test
	void validatePlaceName_nullThrows() {
		assertThatThrownBy(() -> normalizer.validatePlaceName(null))
				.isInstanceOf(PlaceNameValidationException.class)
				.hasMessageContaining("empty");
	}

	@Test
	void validatePlaceName_emptyStringThrows() {
		assertThatThrownBy(() -> normalizer.validatePlaceName(""))
				.isInstanceOf(PlaceNameValidationException.class)
				.hasMessageContaining("empty");
	}

	@Test
	void validatePlaceName_blankStringThrows() {
		assertThatThrownBy(() -> normalizer.validatePlaceName("   "))
				.isInstanceOf(PlaceNameValidationException.class)
				.hasMessageContaining("empty");
	}

	@Test
	void validatePlaceName_201CodePointsThrows() {
		String tooLong = "a".repeat(PlaceNameNormalizer.MAX_CODE_POINTS + 1);
		assertThatThrownBy(() -> normalizer.validatePlaceName(tooLong))
				.isInstanceOf(PlaceNameValidationException.class)
				.hasMessageContaining("200");
	}
}

File: mobile/test/core/place_name_normalizer_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:odm_clarity_woleh_mobile/core/place_name_normalizer.dart';

void main() {
  // ── PLACE_NAMES.md §5 canonical test vectors ─────────────────────────────
  // These three vectors MUST pass on both client and server to guarantee that
  // the normalization pipelines produce identical results.

  group('PLACE_NAMES.md §5 canonical vectors', () {
    test('vector 1 — leading/trailing whitespace + case difference are equal', () {
      expect(normalizePlaceName('  Accra  '), equals(normalizePlaceName('accra')));
    });

    test('vector 2 — decomposed accent equals precomposed after NFC', () {
      // e + U+0301 (combining acute) is NFD for é; U+00E9 is NFC.
      const decomposed = 'e\u0301';
      const precomposed = '\u00e9';
      expect(normalizePlaceName(decomposed), equals(normalizePlaceName(precomposed)));
    });

    test('vector 3 — multiple internal spaces collapse to one', () {
      expect(normalizePlaceName('Main  St'), equals(normalizePlaceName('Main St')));
    });
  });

  // ── normalizePlaceName: edge cases ────────────────────────────────────────

  group('normalizePlaceName edge cases', () {
    test('empty string returns empty', () {
      expect(normalizePlaceName(''), isEmpty);
    });

    test('blank string (spaces only) returns empty', () {
      expect(normalizePlaceName('   '), isEmpty);
    });

    test('internal tab collapsed to single space', () {
      expect(normalizePlaceName('Accra\tCentral'), equals('accra central'));
    });

    test('mixed case is fully folded', () {
      expect(normalizePlaceName('CIRCLE'), equals('circle'));
      expect(normalizePlaceName('Kaneshie'), equals('kaneshie'));
    });

    test('exact output values match expected normal forms', () {
      expect(normalizePlaceName('  Accra  '), equals('accra'));
      expect(normalizePlaceName('Main  St'), equals('main st'));
    });

    test('mixed internal whitespace (tab + spaces) collapses correctly', () {
      expect(normalizePlaceName('Accra \t  Central'), equals('accra central'));
    });

    test('is idempotent — applying twice gives the same result', () {
      const input = '  Kaneshie  Market  ';
      final once = normalizePlaceName(input);
      expect(normalizePlaceName(once), equals(once));
    });

    test('Ghana-English place names normalize to expected forms', () {
      expect(normalizePlaceName('Madina'), equals('madina'));
      expect(normalizePlaceName('MADINA'), equals('madina'));
      expect(normalizePlaceName('  Madina '), equals('madina'));
      expect(normalizePlaceName('Lapaz'), equals('lapaz'));
      expect(normalizePlaceName('Legon'), equals('legon'));
    });
  });

  // ── validatePlaceName ─────────────────────────────────────────────────────

  group('validatePlaceName', () {
    group('valid inputs — returns null', () {
      test('typical place name is accepted', () {
        expect(validatePlaceName('Accra Central'), isNull);
      });

      test('exactly maxPlaceNameCodePoints characters is accepted', () {
        final maxName = 'a' * maxPlaceNameCodePoints;
        expect(validatePlaceName(maxName), isNull);
      });
    });

    group('invalid inputs — returns error string', () {
      test('null returns error', () {
        expect(validatePlaceName(null), isNotNull);
        expect(validatePlaceName(null), contains('empty'));
      });

      test('empty string returns error', () {
        expect(validatePlaceName(''), isNotNull);
        expect(validatePlaceName(''), contains('empty'));
      });

      test('blank string returns error', () {
        expect(validatePlaceName('   '), isNotNull);
        expect(validatePlaceName('   '), contains('empty'));
      });

      test('name exceeding $maxPlaceNameCodePoints code points returns error', () {
        final tooLong = 'a' * (maxPlaceNameCodePoints + 1);
        expect(validatePlaceName(tooLong), isNotNull);
        expect(validatePlaceName(tooLong), contains('$maxPlaceNameCodePoints'));
      });
    });
  });
}
